src/ymfm_adpcm.cpp

```diff
@@ -613,68 +611,101 @@ void adpcm_b_channel::write(uint32_t regnum, uint8_t value)
+			// command byte is written
+			m_curaddress = LATCH_ADDRESS;
+
+			// if playing, set the playing status
+			if (m_regs.execute())
+			{
+				m_buffer = 0;
+				m_nibbles = 0;
+				m_position = 0;
+				m_accumulator = 0;
+				m_adpcm_step = STEP_MIN;
+				m_output = 0;
+
+				set_reset_status(STATUS_PLAYING | STATUS_INTERNAL_PLAYING, STATUS_EOS);
+
+				// don't log masked channels
+				if ((debug::GLOBAL_ADPCM_B_CHANNEL_MASK & 1) != 0)
+					debug::log_keyon("KeyOn ADPCM-B: rep=%d spk=%d pan=%d%d dac=%d 8b=%d rom=%d ext=%d rec=%d start=%04X end=%04X pre=%04X dn=%04X lvl=%02X lim=%04X\n",
+						m_regs.repeat(),
+						m_regs.speaker(),
+						m_regs.pan_left(),
+						m_regs.pan_right(),
+						m_regs.dac_enable(),
+						m_regs.dram_8bit(),
+						m_regs.rom_ram(),
+						m_regs.external(),
+						m_regs.record(),
+						m_regs.start(),
+						m_regs.end(),
+						m_regs.prescale(),
+						m_regs.delta_n(),
+						m_regs.level(),
+						m_regs.limit());
+			}
 		}
-		else
-			m_status &= ~STATUS_EOS;
-		if (m_regs.resetflag())
-			reset();
-		if (m_regs.external())
-			m_dummy_read = 2;
 	}
 
 	// register 8 writes over the bus under some conditions
 	else if (regnum == 0x08)
 	{
-		// if writing from the CPU during execute, clear the ready flag
-		if (m_regs.execute() && !m_regs.record() && !m_regs.external())
-			m_status &= ~STATUS_BRDY;
+		// writing during execute
+		if (m_regs.execute())
+		{
+			// if writing from the CPU during execute, clear the ready flag; data will be picked
+			// up on next fetch
+			if (!m_regs.record() && !m_regs.external())
+				set_reset_status(0, STATUS_BRDY);
+		}
 
-		// if writing during "record", pass through as data
-		else if (!m_regs.execute() && m_regs.record() && m_regs.external())
+		// if writing to external data, process record mode, which writes data to RAM
+		else if (m_regs.external() && m_regs.record())
 		{
-			// clear out dummy reads and set start address
-			if (m_dummy_read != 0)
-			{
-				load_start();
-				m_dummy_read = 0;
-			}
+			// latch the current address if this is the first write
+			if (m_curaddress == LATCH_ADDRESS)
+				latch_addresses();
 
-			// did we hit the end? if so, signal EOS
-			if (at_end())
-			{
-				debug::log_keyon("%s\n", "ADPCM EOS");
-				m_status = STATUS_EOS | STATUS_BRDY;
-			}
+			// write the data
+			m_owner.intf().ymfm_external_write(ACCESS_ADPCM_B, m_curaddress, value);
+			set_reset_status(STATUS_BRDY);
 
-			// otherwise, write the data and signal ready
-			else
+			// advance; if we hit the end, signal EOS and put ourselves back in the latching state
+			// also turn off record mode, as subsequent writes are ignored
+			if (advance_address())
 			{
-				m_owner.intf().ymfm_external_write(ACCESS_ADPCM_B, m_curaddress++, value);
-				m_status = STATUS_BRDY;
+				set_reset_status(STATUS_EOS);
+				m_curaddress = LATCH_ADDRESS;
+				m_regs.write(0x00, m_regs.read(0x00) & ~0x40);
```

Comment:
This behavior is seen only when REPEAT=1.
When REPEAT=0, the second cycle is also a write operation (record mode).

```
if (m_regs.repeat())
    m_regs.write(0x00, m_regs.read(0x00) & ~0x40);
```

Reply:
Hmmm, so if REPEAT=0 then what happens? Does the write continue at the start address? Did you have a test for that case?

Reply:
> Hmmm, so if REPEAT=0 then what happens? Does the write continue at the start address?

Yes.

> Did you have a test for that case?

One of the examples is:
https://github.com/hyano/opna-analyze/blob/main/test/mem_rw.py
https://github.com/hyano/opna-analyze/blob/main/log/mem_rw_mden/mem_rw.txt

I've converted to your C++ framework. Please check it.
[mem_rw.cpp.txt](https://github.com/aaronsgiles/ymfm/files/9159676/mem_rw.cpp.txt)

Currently, I cannot understand actual meaning of REPEAT bit.

Reply:
Ok, I have checked in your rw test. I notice that the previous work to fix dummy reads broke one of the tests, so I made some more changes to fix that. 

I also made another change to initialize memory to 0x80 because the mem_rw test causes a read from uninitialized memory at the end of the "READ ADDRESS 0fff- CHANGE START(10/RESET)" test, so that the dummy reads in the next text "READ ADDRESS 0fff- CHANGE STOP" return that uninitialized memory. Your results showed 0x80 there, so I initialized my memory to 0x80 so that we'd match, but it would be better if the test were fixed.

Reply:
> Your results showed 0x80 there, so I initialized my memory to 0x80 so that we'd match, but it would be better if the test were fixed.

Yes, I also noticed it.
When I create mem_rw.py, dummy read mechanism was unclear and I ignored it.

I have updated my test to fill extended area. And confirmed reading expected values.
https://github.com/hyano/opna-analyze/blob/main/test/mem_rw.py
https://github.com/hyano/opna-analyze/blob/main/log/mem_rw_mden/mem_rw.txt

The log and signals are also recaptured, so you can refer if needed.
https://github.com/hyano/opna-analyze/blob/main/log/mem_rw_mden/

Thank you for using my tests even if it is very ad-hoc and experimental...

[mem_rw.cpp.txt](https://github.com/aaronsgiles/ymfm/files/9172712/mem_rw.cpp.txt)